File: 백준/Silver/14888.연산자 끼워넣기/연산자 끼워넣기.java
import java.util.*;
import java.io.*;


public class Main {
    static int n;
    static int[] arr;
    static int[] oper = new int[4];
    static int max = Integer.MIN_VALUE;
    static int min = Integer.MAX_VALUE;

    static void dfs(int num, int depth){
        if(depth == n){
            max = Math.max(max,num);
            min = Math.min(min,num);
            return;
        }

        for(int i = 0; i < 4; i++){
            if(oper[i] > 0){
                oper[i]--;
                if(i==0){
                    dfs(num+arr[depth],depth+1);
                }else if( i== 1){
                    dfs(num-arr[depth],depth+1);
                }else if( i== 2){
                    dfs(num*arr[depth],depth+1);
                }else if( i== 3){
                    dfs(num/arr[depth],depth+1);
                }

                oper[i]++;
            }
        }

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        arr = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++){
            arr[i] = Integer.parseInt(st.nextToken());
        }
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < 4; i++){
            oper[i] = Integer.parseInt(st.nextToken());
        }
        dfs(arr[0],1);
        System.out.println(max);
        System.out.println(min);
    }
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys

n = int(sys.stdin.readline())

nums = list(map(int, sys.stdin.readline().split()))
oper = list(map(int, sys.stdin.readline().split()))

max_num = -1e9
min_num = 1e9


def dfs(num, depth, add, sub, mul, div):
    global max_num, min_num
    if depth == n:
        max_num = max(max_num, num)
        min_num = min(min_num, num)
        return

    if add > 0:
        dfs(num + nums[depth], depth + 1, add - 1, sub, mul, div)
    if sub > 0:
        dfs(num - nums[depth], depth + 1, add, sub - 1, mul, div)
    if mul > 0:
        dfs(num * nums[depth], depth + 1, add, sub, mul - 1, div)
    if div > 0:
        dfs(int(num / nums[depth]), depth + 1, add, sub, mul, div - 1)


dfs(nums[0], 1, oper[0], oper[1], oper[2], oper[3])
print(max_num)
print(min_num)
